src/csharp/Extensions/Microsoft.Spark.Extensions.DotNet.Interactive/PackageHelper.cs

```diff
@@ -0,0 +1,166 @@
+                    {
+                        FileName = nugetFile.Name,
+                        PackageName = resolvedPackage.PackageName,
+                        PackageVersion = resolvedPackage.PackageVersion
+                    });
+
+                fileAction(nugetFile.FullName);
+            }
+
+            if (nugetMetadata.Count > 0)
+            {
+                var metadataPath =
+                    Path.Combine(
+                        writePath,
+                        DependencyProviderUtils.CreateFileName(++s_metadataCounter));
+                new DependencyProviderUtils.Metadata
+                {
+                    AssemblyProbingPaths = assemblyProbingPaths.ToArray(),
+                    NativeProbingPaths = nativeProbingPaths.ToArray(),
+                    NuGets = nugetMetadata.ToArray()
+                }.Serialize(metadataPath);
+
+                fileAction(metadataPath);
+            }
+        }
+
+        /// <summary>
+        /// In some versions of Spark, spaces is unsupported when using
+        /// <see cref="SparkContext.AddFile(string, bool)"/>.
+        /// 
+        /// For more details please see:
+        /// - https://issues.apache.org/jira/browse/SPARK-30126 
+        /// - https://github.com/apache/spark/pull/26773"
+        /// </summary>
+        /// <param name="path">The path to validate.</param>
+        /// <returns>true if the path is supported by Spark, false otherwise.</returns>
+        internal static bool IsPathValid(string path)
+        {
+            if (!path.Contains(" "))
+            {
+                return true;
+            }
+
+            Version version = SparkEnvironment.SparkVersion;
+            return (version.Major, version.Minor, version.Build) switch
+            {
+                (2, _, _) => false,
+                (3, 0, _) => true,
+                _ => throw new NotSupportedException($"Spark {version} not supported.")
+            };
+        }
+
+        /// <summary>
+        /// Return the delta of the list of packages that have been introduced
+        /// since the last call.
+        /// </summary>
+        /// <returns>The delta of the list of packages.</returns>
+        private static IEnumerable<NuGetPackage> GetPackagesToCopy()
+        {
+            PackageRestoreContext restoreContext =
+                (KernelInvocationContext.Current.HandlingKernel as ISupportNuget)
+                .PackageRestoreContext;
+            IEnumerable<ResolvedPackageReference> packages =
+                restoreContext.ResolvedPackageReferences;
+            foreach (ResolvedPackageReference package in packages)
+            {
+                IEnumerable<FileInfo> files =
+                    package.PackageRoot.EnumerateFiles("*.nupkg", SearchOption.AllDirectories);
+
+                foreach (FileInfo file in files)
+                {
+                    if (!file.Exists || s_filesCopied.Contains(file.Name))
```

Comment:
Would it be easy to debug the `!file.Exists` case if we ever hit it?

Reply:
We can remove this check assuming that we can trust DirectoryInfo.EnumerateFiles(..).